README.md
# react-text-more-less

[![NPM Version](http://img.shields.io/npm/v/react-text-more-less.svg?style=flat-square)](https://www.npmjs.com/package/react-text-more-less)
[![NPM Downloads](https://img.shields.io/npm/dt/react-text-more-less.svg?style=flat-square)](https://www.npmjs.com/package/react-text-more-less)
[![MIT License](https://img.shields.io/npm/l/react-text-more-less.svg?style=flat-square)](./LICENSE)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg?style=flat-square)](https://conventionalcommits.org)

Initialized by [vivaxy/gt-npm-package](https://github.com/vivaxy/gt-npm-package).

Based on [dollarshaveclub/shave](https://github.com/dollarshaveclub/shave).

See [demo](https://vivaxy.github.io/react-text-more-less/demo/index.html).

## Installation

`npm install react-text-more-less`

`yarn add react-text-more-less`

## Usage

```js
import React, { Component } from 'react';
import ReactTextMoreLess from 'react-text-more-less';

class Demo extends Component {
  state = {
    collapsed: true,
  };

  render() {
    const { collapsed } = this.state;
    return (
      <ReactTextMoreLess
        collapsed={collapsed}
        text="1926年1—6月，他一连发表了四篇论文，题目都是《量子化就是本征值问题》，系统地阐明了波动力学理论。在此以前，德国物理学家W.K.海森堡、M.玻恩和E.P.约旦于1925年7—9月通过另一途径建立了矩阵力学。1926年3月，薛定谔发现波动力学和矩阵力学在数学上是等价的，是量子力学的两种形式，可以通过数学变换,从一个理论转到另一个理论。薛定谔起初试图把波函数解释为三维空间中的振动，把振幅解释为电荷密度，把粒子解释为波包。但他无法解决“波包扩散”的困难。最后物理学界普遍接受了玻恩提出的波函数的几率解释。"
        lessHeight={72}
        showMoreText="... 展示更多"
        showMoreElement={
          <span>
            ... <span className="show-more-text">展示更多</span>
          </span>
        }
        showLessElement={<span className="show-more-text">收起</span>}
        onClick={() => {
          this.setState({ collapsed: !collapsed });
        }}
      />
    );
  }
}
```

If you like to prevent collapse on user select, add this before setting `collapsed`.

```js
const selection = window.getSelection();
if (!selection.toString()) {
  /* Set `collapsed` prop */
}
```

## Props

| name              | type    | isRequired | default   | description                                              |
| ----------------- | ------- | ---------- | --------- | -------------------------------------------------------- |
| `text`            | string  | ✔          | N/A       | text in the container                                    |
| `lessHeight`      | number  | ✔          | N/A       | container collapsed height                               |
| `collapsed`       | bool    | ✖          | true      | is the container collapsed according to the `lessHeight` |
| `className`       | string  | ✖          | undefined | className of the container                               |
| `showMoreText`    | string  | ✖          | ...       | the ellipse text to show more                            |
| `showMoreElement` | element | ✖          | undefined | the element to show when container collapsed             |
| `showLessElement` | element | ✖          | undefined | the element to show when container expanded              |
| `onClick`         | func    | ✖          | () => {}  | container click handler                                  |
| `rootProps`       | object  | ✖          | {}        | root container props, could be any props a `div` accepts |

If `showMoreElement` supplied, `showMoreText` will be replaced, this props is useful when you want to custom `showMoreText` style or to use a more complex dom structure to display `showMoreText`.

# Related Projects

- [react-ellipsis-text](https://github.com/mobilusoss/react-ellipsis-text)
- [react-truncate](https://github.com/pablosichert/react-truncate)
- [react-text-truncate](https://github.com/ShinyChang/React-Text-Truncate)
- [shiitake](https://github.com/bsidelinger912/shiitake)
- [react-read-more-read-less](https://github.com/giatro/react-read-more-read-less)
- [react-lines-ellipsis](https://github.com/xiaody/react-lines-ellipsis)
- [react-clamp-lines](https://github.com/zoltantothcom/react-clamp-lines)
- [read-more-react](https://github.com/alexandersmanning/read-more-react)

# Change Log

[Change Log](./CHANGELOG.md)

# Licence

[MIT](./LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.4.2](https://github.com/vivaxy/react-text-more-less/compare/v0.4.1...v0.4.2) (2020-11-11)


### Bug Fixes

* :bug: fix a problem when toggle collapse not working ([465cf25](https://github.com/vivaxy/react-text-more-less/commit/465cf259c078e1b3135dabfedef2461118e0e893))

### [0.4.1](https://github.com/vivaxy/react-text-more-less/compare/v0.4.0...v0.4.1) (2020-06-07)

<a name="0.4.0"></a>

# [0.4.0](https://github.com/vivaxy/react-text-more-less/compare/v0.3.0...v0.4.0) (2018-08-22)

### Features

- **dependency upgrade:** :arrow_up:shave[@2](https://github.com/2).4.0 ([dc1fd8d](https://github.com/vivaxy/react-text-more-less/commit/dc1fd8d))

<a name="0.3.0"></a>

# [0.3.0](https://github.com/vivaxy/react-text-more-less/compare/v0.2.0...v0.3.0) (2017-08-19)

### Features

- **props:** :sparkles:new props: `rootProps` ([88c7245](https://github.com/vivaxy/react-text-more-less/commit/88c7245))

<a name="0.2.0"></a>

# [0.2.0](https://github.com/vivaxy/react-text-more-less/compare/v0.1.0...v0.2.0) (2017-08-19)

### Features

- **keywords:** :sparkles: ([ac3c731](https://github.com/vivaxy/react-text-more-less/commit/ac3c731))

<a name="0.1.0"></a>

# 0.1.0 (2017-08-12)

### Bug Fixes

- **demo:** :bug:Fix demo ([eec21e9](https://github.com/vivaxy/react-text-more-less/commit/eec21e9))
- **demo:** :bug:Fix demo ([b9f08c9](https://github.com/vivaxy/react-text-more-less/commit/b9f08c9))

### Features

- :sparkles: ([4c17f7f](https://github.com/vivaxy/react-text-more-less/commit/4c17f7f))


src/index.js
/**
 * @since 2017-04-11 19:10:08
 * @author vivaxy
 */
import React, { Component } from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import PropTypes from 'prop-types';
import shave from 'shave';

const showLessTextClassName = 'js-show-less-text';
const jsShaveCharClassName = 'js-shave-char';
const resizeEventName = 'resize';

export default class ReactTextMoreLess extends Component {
  static propTypes = {
    text: PropTypes.string.isRequired,
    lessHeight: PropTypes.number.isRequired,
    collapsed: PropTypes.bool,
    className: PropTypes.string,
    showMoreText: PropTypes.string,
    showMoreElement: PropTypes.element,
    showLessElement: PropTypes.element,
    onClick: PropTypes.func,
    rootProps: PropTypes.object,
  };

  static defaultProps = {
    collapsed: true,
    className: undefined,
    showMoreText: '...',
    showMoreElement: undefined,
    showLessElement: undefined,
    onClick: () => {},
    rootProps: {},
  };

  componentDidMount() {
    window.addEventListener(resizeEventName, this.setDOM);
    this.setDOM();
  }

  getSnapshotBeforeUpdate(prevProps) {
    // Reset the DOM before update, this makes React render properly.
    this.resetDOM(prevProps);
    return null;
  }

  componentDidUpdate() {
    this.setDOM();
  }

  componentWillUnmount() {
    this.resetDOM(this.props);
    window.removeEventListener(resizeEventName, this.setDOM);
  }

  setRootRef = (root) => {
    this.root = root;
  };

  setDOM = () => {
    const {
      className,
      collapsed,
      showMoreText,
      lessHeight,
      showMoreElement,
      showLessElement,
    } = this.props;
    if (collapsed) {
      shave(this.root, lessHeight, {
        classname: className,
        character: showMoreText,
        spaces: false,
      });
      const shaveChar = this.root.querySelector(`.${jsShaveCharClassName}`);
      if (shaveChar && showMoreElement) {
        // 如果不需要 ellipse，则没有 `shaveChar`
        render(showMoreElement, shaveChar);
      }
    } else {
      const hasShowLess = !!this.root.querySelector(
        `.${showLessTextClassName}`
      );
      if (
        this.root.offsetHeight > lessHeight &&
        showLessElement &&
        !hasShowLess
      ) {
        // 如果需要收起，则展示文字
        const collapse = document.createElement('span');
        collapse.setAttribute('class', showLessTextClassName);
        render(showLessElement, collapse);
        this.root.appendChild(collapse);
      }
    }
  };

  resetDOM(prevProps) {
    // clean up everything
    const { collapsed, showMoreElement } = prevProps;
    if (collapsed) {
      const shaveChar = this.root.querySelector(`.${jsShaveCharClassName}`);
      if (shaveChar && showMoreElement) {
        unmountComponentAtNode(shaveChar);
      }
      this.root.innerHTML = this.props.text;
    } else {
      const collapse = this.root.querySelector(`.${showLessTextClassName}`);
      if (collapse) {
        unmountComponentAtNode(collapse);
        this.root.removeChild(collapse);
      }
    }
  }

  render() {
    const { className, text, onClick, rootProps } = this.props;
    return (
      <div
        ref={this.setRootRef}
        className={className}
        onClick={onClick}
        {...rootProps}
      >
        {text}
      </div>
    );
  }
}
